### PR TITLE
[TS] Refactor TS module

### PR DESCRIPTION
**Pull request by nlordell**
This refactors the main module so it is now a TS module. This means that TS projects (like dex-price-estimator) should integrate better with the contracts as it is now a TS module.

### Test Plan

Use `yarn link` with a TS project and ensure that this new module system works as intended. I will add detailed instructions soon.

### CODE REVIEW

**Review by bh2smith — APPROVED** [review]
This is good, but the changes seem like the content of two PRs.

**Comment by nlordell** [comment]
> This is good, but the changes seem like the content of two PRs.

Do you mean the `@src` path that was added should be a separate PR? I can do that.

**edit**: @bh2smith I reduced it as much as possible while keeping it buildable with the new structure. Let me know if its easier to review :).

File: package.json [hunk 1]
```diff
@@ -2,10 +2,10 @@
   "name": "@gnosis.pm/dex-contracts",
   "version": "0.2.15",
   "description": "Contracts for dFusion multi-token batch auction exchange",
-  "main": "src/index.js",
+  "main": "build/commonjs/index.ts",
```
Comment (fleupold) [inline package.json]: Isn't the path just common and doesn't the file still end in js?

File: src/encoding.d.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import BN from "bn.js";
```
Comment (fleupold) [inline src/encoding.d.ts]: Will this conflict with #753 ?

**Review by fleupold — APPROVED** [review]
Seems fine. Let's coordinate on how to merge this with the other outstanding ts paths

File: src/encoding.d.ts (hunk 2 above)
Comment (nlordell) [reply src/encoding.d.ts]: You should be able to just remove the `encoding.d.ts` file once `encoding.js` becomes `encoding.ts`.

File: package.json (hunk 1 above)
Comment (nlordell) [reply package.json]: The path was wrong, I noticed it when I used `yarn link` with `dex-price-estimator`.

**Comment by nlordell** [comment]
> Seems fine. Let's coordinate on how to merge this with the other outstanding ts paths

Looks like `tsc` doesn't copy .js files to the build output, so the import paths on the compiled JS files for `encoding.js` and `onchain_reading.js` are currently wrong. This means for this to work correctly, we need to merge #753 and #754 first anyway.

File: tsconfig.json [hunk 3]
```diff
@@ -8,7 +8,7 @@
     // "allowJs": true,                       /* Allow javascript files to be compiled. */
     // "checkJs": true,                       /* Report errors in .js files. */
     // "jsx": "preserve",                     /* Specify JSX code generation: 'preserve', 'react-native', or 'react'. */
-    // "declaration": true,                   /* Generates corresponding '.d.ts' file. */
+    "declaration": true,                   /* Generates corresponding '.d.ts' file. */
```
Comment (bh2smith) [inline tsconfig.json:11]: What are these `.d.ts` files anyway?

Comment (nlordell) [reply tsconfig.json:11]: Type definitions. They allow other TypeScript projects to use type information from the original TS code.

**Pull request merged**